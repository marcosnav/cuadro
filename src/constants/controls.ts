/**
 * Available control types
 */
export enum ControlType {
  NEW_GAME = 'NEW_GAME',
  RESTART_GAME = 'RESTART_GAME',
  SEE_ORIGINAL = 'SEE_ORIGINAL',
}
