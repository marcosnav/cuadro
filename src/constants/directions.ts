/**
 * SwipeDirection, one of UP, RIGHT, LEFT, DOWN
 */
export enum SwipeDirection {
  UP = 'UP',
  RIGHT = 'RIGHT',
  DOWN = 'DOWN',
  LEFT = 'LEFT',
}
