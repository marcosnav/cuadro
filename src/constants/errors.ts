export enum errors {
  NO_INTERNET = 'NO_INTERNET',
  UNEXPECTED_ERROR = 'UNEXPECTED_ERROR',
  UNRESPONSIVE_IMAGE_PROVIDER = 'UNRESPONSIVE_IMAGE_PROVIDER',
}
