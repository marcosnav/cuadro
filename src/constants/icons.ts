import check from './../assets/svg/check.svg';
import image from './../assets/svg/image.svg';
import newicon from './../assets/svg/new.svg';
import refresh from './../assets/svg/refresh.svg';

export const icons: { [key: string]: string } = {
  check,
  image,
  new: newicon,
  refresh,
};
