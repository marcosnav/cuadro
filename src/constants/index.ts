export * from './status';
export * from './errors';
export * from './theme';
export * from './keycodes';
