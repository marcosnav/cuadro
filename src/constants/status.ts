export enum status {
  LOADING_IMAGE = 'LOADING_IMAGE',
  DISPLAY_ORIGINAL = 'DISPLAY_ORIGINAL',
  PLAYING_GAME  = 'PLAYING_GAME ',
  FINISHED_GAME = 'FINISHED_GAME',
}
