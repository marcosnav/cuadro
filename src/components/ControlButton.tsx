import React, { FC } from 'react';
import styled from 'styled-components';
import { icons } from './../constants';

const ButtonWrapper = styled.button`
  align-items: center;
  background: ${(props) => props.theme.WHITE};
  border: none;
  color: ${(props) => props.theme.DARK};
  display: flex;
  flex-direction: column;
  font-size: 0.64em;
  font-weight: bold;
  height 42px;
  justify-content: space-between;
  opacity: ${(props) => props.disabled ? '0.3' : '1'};
  padding: 0;

  &:focus {
    outline: none;
  }
`;

const ButtonIcon = styled.img`
  height: 26px;
  width: auto;
`;

/**
 * Available control types
 */
export enum ControlType {
  NEW_GAME = 'NEW_GAME',
  RESTART_GAME = 'RESTART_GAME',
  SEE_ORIGINAL = 'SEE_ORIGINAL',
}

interface IProps {
  type?: ControlType;
  disabled?: boolean;
  action: () => void;
}

/**
 * ControlButton Component
 * @param props { type: ControlType, disabled: boolean }
 */
const ControlButton: FC<IProps> = (props) => {
  const { action, type, disabled } = props;
  let icon: string = '';

  switch (type) {
    case ControlType.NEW_GAME:
      icon = icons.new;
      break;
    case ControlType.RESTART_GAME:
      icon = icons.refresh;
      break;
    case ControlType.SEE_ORIGINAL:
      icon = icons.image;
  }

  const evalAction = () => {
    if (disabled) {
      return;
    }
    action();
  };

  return (
    <ButtonWrapper onClick={evalAction} disabled={disabled}>
      <ButtonIcon src={icon} />
      {props.children}
    </ButtonWrapper>
  );
};

ControlButton.defaultProps = {
  disabled: false,
};

export default ControlButton;
