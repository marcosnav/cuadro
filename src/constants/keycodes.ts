import { SwipeDirection } from './directions';

/**
 * KeyCodes to match for keypress events
 */
export enum KeyCodes {
  DOWN = 40,
  LEFT = 37,
  RIGHT = 39,
  UP = 38,
  ENTER = 13,
}

/**
 * Map key code with swipe direction
 */
export const KeySwipeMap: { [key: number]: SwipeDirection } = {
  40: SwipeDirection.DOWN,
  37: SwipeDirection.LEFT,
  39: SwipeDirection.RIGHT,
  38: SwipeDirection.UP,
};
