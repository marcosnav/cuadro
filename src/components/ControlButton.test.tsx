import { mount } from 'enzyme';
import 'jest-enzyme';
import 'jest-styled-components';
import React from 'react';
import { icons } from './../constants';
import ControlButton,  { ControlType } from './ControlButton';

describe('<ControlButton />', () => {
  test('renders properly', () => {
    const actionSpy = jest.fn();
    const component = mount(<ControlButton action={actionSpy} >My Control Button</ControlButton>);
    component.simulate('click');
    expect(component).toMatchSnapshot();
    expect(actionSpy).toHaveBeenCalled();
  });

  describe('when disabled', () => {
    const actionSpy = jest.fn();
    const component = mount(<ControlButton action={actionSpy} disabled={true} >disabled</ControlButton>);

    test('click wont work and must be opaque', () => {
      component.simulate('click');
      expect(component).toHaveStyleRule('opacity', '0.3');
      expect(component.text()).toBe('disabled');
      expect(actionSpy).not.toHaveBeenCalled();
    });

    test('when re-enabled, click must work and be full color', () => {
      component.setProps({ disabled: false });
      component.simulate('click');
      expect(component).toHaveStyleRule('opacity', '1');
      expect(actionSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('with ControlType.NEW_GAME', () => {
    test('the icon must be the plus/new/add icon', () => {
      const actionSpy = jest.fn();
      const component = mount(<ControlButton action={actionSpy} type={ControlType.NEW_GAME}>new</ControlButton>);
      const icon = component.find('img');
      component.simulate('click');
      expect(component.text()).toBe('new');
      expect(actionSpy).toHaveBeenCalled();
      expect(icon.getElement().props.src).toBe(icons.new);
    });
  });

  describe('with ControlType.RESTART_GAME', () => {
    test('the icon must be the reload/restart icon', () => {
      const actionSpy = jest.fn();
      const component = mount(
      <ControlButton action={actionSpy} type={ControlType.RESTART_GAME} >restart</ControlButton>);
      const icon = component.find('img');
      component.simulate('click');
      expect(component.text()).toBe('restart');
      expect(actionSpy).toHaveBeenCalled();
      expect(icon.getElement().props.src).toBe(icons.refresh);
    });
  });

  describe('with ControlType.SEE_ORIGINAL', () => {
    test('the icon must be the image icon', () => {
      const actionSpy = jest.fn();
      const component = mount(
      <ControlButton action={actionSpy} type={ControlType.SEE_ORIGINAL}>see original</ControlButton>);
      const icon = component.find('img');
      component.simulate('click');
      expect(component.text()).toBe('see original');
      expect(actionSpy).toHaveBeenCalled();
      expect(icon.getElement().props.src).toBe(icons.image);
    });
  });
});
