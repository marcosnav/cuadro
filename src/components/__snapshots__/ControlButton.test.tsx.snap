// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControlButton /> renders properly 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 0.64em;
  font-weight: bold;
  height: 42px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  opacity: 1;
  padding: 0;
}

.c0:focus {
  outline: none;
}

.c1 {
  height: 26px;
  width: auto;
}

<ControlButton
  action={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  disabled={false}
>
  <styled.button
    disabled={false}
    onClick={[Function]}
  >
    <StyledComponent
      disabled={false}
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  align-items: center;
  background: ",
              [Function],
              ";
  border: none;
  color: ",
              [Function],
              ";
  display: flex;
  flex-direction: column;
  font-size: 0.64em;
  font-weight: bold;
  height 42px;
  justify-content: space-between;
  opacity: ",
              [Function],
              ";
  padding: 0;

  &:focus {
    outline: none;
  }
",
            ],
          },
          "displayName": "styled.button",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onClick={[Function]}
    >
      <button
        className="c0"
        disabled={false}
        onClick={[Function]}
      >
        <styled.img
          src=""
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": true,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  height: 26px;
  width: auto;
",
                  ],
                },
                "displayName": "styled.img",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "img",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            src=""
          >
            <img
              className="c1"
              src=""
            />
          </StyledComponent>
        </styled.img>
        My Control Button
      </button>
    </StyledComponent>
  </styled.button>
</ControlButton>
`;
